src/sequence_batch_scheduler.cc

```diff
@@ -110,12 +108,143 @@ SequenceBatchScheduler::Create(
         config.sequence_batching().oldest().max_candidate_sequences();
   }
 
+  // Create batchers.
+  RETURN_IF_ERROR(sched->CreateBatchers(model->BackgroundInstances()));
+
+  // Create a reaper thread that watches for idle sequences. Run the
+  // reaper a lower priority.
+  SequenceBatchScheduler* raw = sched.release();
+
+  raw->reaper_thread_exit_ = false;
+  raw->reaper_thread_.reset(
+      new std::thread([raw]() { raw->ReaperThread(10 /* nice */); }));
+
+  scheduler->reset(raw);
+
+  return Status::Success;
+}
+
+Status
+SequenceBatchScheduler::Update()
+{
+  // Find added and removed instances for this update.
+  std::vector<std::shared_ptr<TritonModelInstance>> added_instances;
+  std::unordered_set<TritonModelInstance*> removed_instances;
+  InstancesDiff(&added_instances, &removed_instances);
+
+  // Batchers should be destroyed outside the lock.
+  std::vector<std::unique_ptr<SequenceBatch>> removed_batchers;
+
+  {
+    std::unique_lock<std::mutex> lk(mu_);
+
+    // Add the new batchers.
+    RETURN_IF_ERROR(CreateBatchers(added_instances));
+
+    // All sequence slots of the removed instances are pending to be removed.
+    pending_removal_seq_slots_.clear();
+    for (auto& instance : removed_instances) {
+      pending_removal_seq_slots_.emplace(instance, seq_slot_cnt_);
+    }
+
+    // Erase ready slots which its instance is pending removal.
+    std::priority_queue<
+        BatcherSequenceSlot, std::vector<BatcherSequenceSlot>,
+        BatcherSequenceSlotCompare>
+        new_ready_batcher_seq_slots;
+    while (!ready_batcher_seq_slots_.empty()) {
+      auto& ready_batcher_seq_slot = ready_batcher_seq_slots_.top();
+      if (removed_instances.find(ready_batcher_seq_slot.model_instance_) ==
+          removed_instances.end()) {
+        // The ready slot is not being removed.
+        new_ready_batcher_seq_slots.push(ready_batcher_seq_slot);
+      } else {
+        // The ready slot is removed, so update pending removal instances.
+        ErasePendingRemovalSequenceSlot(ready_batcher_seq_slot);
+      }
+      ready_batcher_seq_slots_.pop();
+    }
+    ready_batcher_seq_slots_.swap(new_ready_batcher_seq_slots);
+
+    // Wait until all removed batchers have completed its assigned sequence.
+    pending_removal_seq_slots_cv_.wait(
+        lk, [this] { return pending_removal_seq_slots_.empty(); });
+
+    // Erase all removed batchers, outside the lock.
+    for (auto& instance : removed_instances) {
+      removed_batchers.emplace(removed_batchers.end(), nullptr);
+      auto it = batchers_.find(instance);
+      it->second.swap(removed_batchers.back());
+      batchers_.erase(std::move(it));
+    }
+
+    // Update removed request count.
+    for (auto& instance : removed_instances) {
+      queue_request_cnts_.erase(instance);
+    }
+  }
+
+  return Status::Success;
+}
+
+void
+SequenceBatchScheduler::InstancesDiff(
+    std::vector<std::shared_ptr<TritonModelInstance>>* added_instances,
```

Comment:
could use references here?

Reply:
Yes, but I think @GuanLuo mentioned to me that we want to use pointers for parameters that will be modified by the function (i.e. [FindModelIdentifier](https://github.com/triton-inference-server/core/blob/main/src/model_repository_manager.h#L469-L470)). Unless this does not apply on the model and scheduler part of the code.

Reply:
I would just keep consistent with other functions - so sounds good to me. We can reconsider later.